ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,193 @@
+   * @param requestsToDrop a list of correlation IDs that will contain the IDs for requests that the network layer
+   *                       should drop.
+   */
+  private void pollNewRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    putManager.poll(requestsToSend, requestsToDrop);
+    getManager.poll(requestsToSend, requestsToDrop);
+    nonBlockingRouter.initiateBackgroundDeletes(getBackGroundDeleteRequests());
+    clearBackGroundDeleteRequests();
+    deleteManager.poll(requestsToSend, requestsToDrop);
+    ttlUpdateManager.poll(requestsToSend, requestsToDrop);
+    undeleteManager.poll(requestsToSend, requestsToDrop);
+  }
+
+  /**
+   * Add the specified {@link List} of {@link RequestInfo}s to the reqeust queue.
+   * @param requestInfos {@link List} of {@link RequestInfo} objects.
+   */
+  private void addToRequestQueue(List<RequestInfo> requestInfos) {
+    for (RequestInfo requestInfo : requestInfos) {
+      QuotaResource quotaResource = requestInfo.getChargeable().getQuotaResource();
+      if (quotaResource == null) {
+        quotaResource = UNKNOWN_QUOTA_RESOURCE;
+      }
+      requestQueue.putIfAbsent(quotaResource, new LinkedList<>());
+      requestQueue.get(quotaResource).add(requestInfo);
+    }
+  }
+
+  /**
+   * Drain the request queue based on quota and system resources and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainRequestQueue(List<RequestInfo> requestsToSend) {
+    drainQuotaCompliantRequests(requestsToSend);
+
+    List<QuotaResource> quotaResources = new ArrayList<>(requestQueue.keySet());
+    Collections.shuffle(quotaResources);
+    boolean done = false;
+    while (!requestQueue.isEmpty()) {
+      for (QuotaResource quotaResource : quotaResources) {
+        RequestInfo requestInfo = requestQueue.get(quotaResource).getFirst();
+        if (requestInfo.getChargeable().quotaExceedAllowed()) {
+          requestInfo.getChargeable().charge();
+          requestsToSend.add(requestInfo);
+          requestQueue.get(quotaResource).removeFirst();
+          if (requestQueue.get(quotaResource).isEmpty()) {
+            requestQueue.remove(quotaResource);
+          }
+        } else {
+          done = true;
+          break;
+        }
+      }
+      if (done) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Drain the request queue based on resource quota only and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainQuotaCompliantRequests(List<RequestInfo> requestsToSend) {
+    for (QuotaResource quotaResource : requestQueue.keySet()) {
+      if (quotaResource.equals(UNKNOWN_QUOTA_RESOURCE)) {
+        // If there are requests for which QuotaResource couldn't be found, this is most likely a bug.
```

Comment:
Do you want to add some metrcis for this case?

Reply:
I have a separate ticket for metrics and alerting. Will share the jira ticket with you guys separately.